lottery/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+}
+
+impl Lottery {
+    // checks that lottery has started and lottery time has not expired
+    fn lottery_is_on(&mut self) -> bool {
+        self.lottery_state.lottery_started
+            && (self.lottery_state.lottery_start_time + self.lottery_state.lottery_duration)
+                > exec::block_timestamp()
+    }
+
+    /// Launches a lottery
+    /// Requirements:
+    /// * Only owner can launch lottery
+    /// * Lottery must not have been launched earlier
+    /// Arguments:
+    /// * `duration`: lottery duration in milliseconds
+    /// * `token_address`: address of Fungible Token contract
+    fn start_lottery(&mut self, duration: u64, token_address: Option<ActorId>) {
+        if msg::source() == self.lottery_owner && !self.lottery_is_on() {
+            self.lottery_state.lottery_started = true;
+            self.lottery_state.lottery_start_time = exec::block_timestamp();
+            self.lottery_state.lottery_duration = duration;
+            self.token_address = token_address;
+            self.lottery_id += 1;
+            self.lottery_balance = 0;
+        } else {
+            panic!(
+                "start_lottery(): Lottery on: {}  Owner message: {}",
+                self.lottery_is_on(),
+                msg::source() == self.lottery_owner
+            );
+        }
+    }
+
+    // checks that the player is already participating in the lottery
+    fn player_exists(&mut self) -> bool {
+        self.players
+            .values()
+            .any(|player| player.player_id == msg::source())
+    }
+
+    /// Transfers `amount` tokens from `sender` account to `recipient` account.
+    /// Arguments:
+    /// * `from`: sender account
+    /// * `to`: recipient account
+    /// * `amount`: amount of tokens
+    async fn transfer_tokens(&mut self, from: &ActorId, to: &ActorId, amount_tokens: u128) {
+        let _transfer_response: Event = msg::send_and_wait_for_reply(
+            self.token_address.unwrap(),
+            Action::Transfer {
+                from: *from,
+                to: *to,
+                amount: amount_tokens,
+            },
+            0,
+        )
+        .await
+        .expect("Error in transfer");
+    }
+
+    /// Called by a player in order to participate in lottery
+    /// Requirements:
+    /// * Lottery must be on
+    /// * Contribution must be greater than zero
+    /// * The player cannot enter the lottery more than once
+    /// Arguments:
+    /// * `amount`: value or tokens to participate in the lottery
+    async fn enter(&mut self, amount: u128) {
+        if self.lottery_is_on() && !self.player_exists() && amount > 0 {
+            let player = Player {
+                player_id: msg::source(),
+                balance: amount,
+            };
+
+            if self.token_address.is_some() {
+                self.transfer_tokens(&msg::source(), &exec::program_id(), amount)
+                    .await;
+
+                self.lottery_balance += amount;
```

Comment:
It is a redundant action because the lottery balance is counted in the fungible token contract.